ofrak_core/ofrak/service/new_data_service.py

```diff
@@ -0,0 +1,409 @@
+            else:
+                patches_by_root[target_data_model.id].append(patch)
+
+        results = []
+        for root_id, patches_for_root in patches_by_root.items():
+            results.extend(self._apply_patches_to_root(root_id, patches_for_root, []))
+
+        return results
+
+    async def delete_node(self, data_id: bytes) -> None:
+        model = self._get_by_id(data_id)
+        if model.root_id is None:
+            # deleting a root node
+            root = self._roots[model.id]
+            for child_model in root.get_children():
+                del self._model_store[child_model.id]
+
+            del self._roots[model.id]
+
+        else:
+            root = self._get_root_by_id(model.root_id)
+            root.delete_model(model)
+            del self._model_store[data_id]
+
+    async def delete_models(self, data_ids: Iterable[bytes]) -> None:
+        roots_to_delete = set()
+        mapped_to_delete = set()
+
+        for data_id in data_ids:
+            model = self._get_by_id(data_id)
+            if model.root_id is None:
+                roots_to_delete.add(model)
+            else:
+                mapped_to_delete.add(model)
+
+        for root_model in roots_to_delete:
+            root = self._roots[root_model.id]
+            for child_model in root.get_children():
+                mapped_to_delete.remove(child_model)
+                del self._model_store[child_model.id]
+
+            del self._roots[root_model.id]
+
+        for model in mapped_to_delete:
+            root = self._get_root_by_id(model.root_id)
+            root.delete_model(model)
+            del self._model_store[model.id]
+
+    def _apply_patches_to_root(
+        self,
+        root_data_id: DataId,
+        patches: List[DataPatch],
+        moves: List[DataMove],
+    ) -> List[DataPatchesResult]:
+        root: _DataRoot = self._roots[root_data_id]
+        finalized_ordered_patches: List[Tuple[Range, bytes]] = []
+        resize_tracker = _PatchResizeTracker()
+
+        # Screen patches for inconsistencies/overlaps
+        # And translate them as required by preceding patches
+        raw_patch_ranges_in_root = []
+
+        for patch in patches:
+            if resize_tracker.overlaps_resized_range(patch.range):
+                raise PatchOverlapError(
+                    f"Patch to {patch.range} of {patch.data_id.hex()} overlaps previously resized "
+                    f"area of {patch.data_id.hex()} and cannot be applied!"
+                )
+            target_model = self._get_by_id(patch.data_id)
+
+            raw_absolute_patch_range = self._get_absolute_range(target_model, patch.range)
+            raw_patch_ranges_in_root.append(raw_absolute_patch_range)
+            absolute_patch_range = resize_tracker.translate_range(raw_absolute_patch_range)
+            models_intersecting_patch_range = root.get_children_with_boundaries_intersecting_range(
+                absolute_patch_range
+            )
```

Comment:
What is the difference between a `raw_absolute_patch_range` and an `absolute_patch_range`?

Reply:
`absolute_patch_range` is translated according to any shifts in the data that may have happened as a result of earlier patches. For example, Path 1 resizes range (0,4) to have length 2 (size diff = -2), then a patch to range (6, 10) should be shifted to be applied to range (4, 8)

Reply:
So if I understand correctly:

- `raw_absolute_patch_range` is the range of the patch in the root data
- `absolute_patch_range` is the translated range if any shifts have occurred

What about renaming:
- `_get_absolute_range` -> `_get_range_in_root_data`
- `raw_absolute_patch_range -> `patch_range_in_root`
- `absolute_patch_range` -> `translated_patch_range_in_root` (or `shifted_...` ?)